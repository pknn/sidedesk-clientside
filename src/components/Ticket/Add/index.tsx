import React from 'react'
import { FontAwesomeIcon as Icon } from '@fortawesome/react-fontawesome'
import { faPlus } from '@fortawesome/free-solid-svg-icons'

export const Add = () => (
  <div className="m-2 p-2 flex items-center hover:bg-gray-300 rounded cursor-pointer">
    <Icon icon={faPlus} size="xs"></Icon>
    <span className="text-sm ml-2">Add item</span>
  </div>
)
