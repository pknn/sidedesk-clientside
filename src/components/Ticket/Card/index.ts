export { CardComponent as Card } from './CardComponent'
