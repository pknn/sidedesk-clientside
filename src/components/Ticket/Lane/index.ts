export { LaneComponent as Lane } from './LaneComponent'
