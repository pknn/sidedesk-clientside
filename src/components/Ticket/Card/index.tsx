import React from 'react'
import {
  Draggable,
  DraggableProvided,
  DraggableStateSnapshot,
} from 'react-beautiful-dnd'

import { Ticket } from 'app/types/Ticket'
import { CardComponent } from './CardComponent'
import { useAppDispatch } from 'app/types/Store'
import { actions } from 'app/store/features/Metadata/slice'

interface DataProps {
  ticket: Ticket
  index: number
}

export const Card = ({ ticket, index }: DataProps): JSX.Element => {
  const dispatch = useAppDispatch()

  const handleTicketClick = () => {
    dispatch(actions.showEditTicketModal({ withTicket: ticket }))
  }

  return (
    <Draggable draggableId={ticket.id.toString()} key={ticket.id} index={index}>
      {(
        draggableProvided: DraggableProvided,
        draggableStateSnapshot: DraggableStateSnapshot,
      ) => (
        <CardComponent
          id={ticket.id}
          title={ticket.title}
          reporterName={ticket.reporterName}
          draggableProvided={draggableProvided}
          draggableStateSnapshot={draggableStateSnapshot}
          onClick={handleTicketClick}
        />
      )}
    </Draggable>
  )
}
