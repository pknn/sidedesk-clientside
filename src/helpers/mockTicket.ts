import faker from 'faker'

import { Ticket, TicketForm, TicketStatus } from 'app/types/Ticket'

export const ticketStatusOptions: TicketStatus[] = [
  TicketStatus.Pending,
  TicketStatus.Accepted,
  TicketStatus.Resolved,
  TicketStatus.Rejected,
]

type DateIndex = 0 | 1 | 2

const dates: Date[] = [
  new Date(1623862800),
  new Date(1623949200),
  new Date(1624035600),
]

const getFakeDate = (dateIndex: DateIndex): Date => dates[dateIndex]

const getFakeTicket = (
  status?: TicketStatus,
  dateIndex?: DateIndex,
): Ticket => {
  const firstName = faker.name.firstName()
  const lastName = faker.name.lastName()
  const fullName = `${firstName} ${lastName}`
  return {
    id: faker.datatype.number(),
    title: faker.lorem.sentence(10),
    description: faker.lorem.sentences(4),
    reporterName: fullName,
    reporterEmail: faker.internet.email(firstName, lastName),
    status: status || faker.random.arrayElement(ticketStatusOptions),
    createdAt: getFakeDate(dateIndex || 0),
    updatedAt: getFakeDate(dateIndex || 0),
  }
}

export const getMockTicket = (status?: TicketStatus): Ticket => {
  faker.seed(96421)
  return getFakeTicket(status)
}

export const getMockTickets = (n: number, status?: TicketStatus): Ticket[] => {
  faker.seed(96421)
  const tickets: Ticket[] = []
  for (let i = 0; i < n; i += 1) {
    const fa = i % 3 == 0
    const fb = i % 3 == 1
    tickets.push(getFakeTicket(status, fa ? 0 : fb ? 1 : 2))
  }
  return tickets
}

export const getTickets = (n: number, status: TicketStatus): Ticket[] => {
  const tickets: Ticket[] = []
  for (let i = 0; i < n; i += 1) {
    const fa = i % 3 == 0
    const fb = i % 3 == 1
    tickets.push(getFakeTicket(status, fa ? 0 : fb ? 1 : 2))
  }
  return tickets
}
