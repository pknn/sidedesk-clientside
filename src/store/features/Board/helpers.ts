import { DraggableLocation } from 'react-beautiful-dnd'

import {
  MoveCrossLaneActionPayload,
  MoveInLaneActionPayload,
  TicketPosition,
} from 'app/store/features/Board/types'
import { Ticket } from 'app/types/Ticket'
import {
  getStatusFromKey,
  getStatusFromString,
  TicketStatusKeys,
} from 'app/helpers/statusMappers'

export const getMoveInLaneActionPayload = (
  source: DraggableLocation,
  destination: DraggableLocation,
): MoveInLaneActionPayload => ({
  laneId: source.droppableId,
  from: source.index,
  to: destination.index,
})

export const getMoveCrossLaneActionPayload = (
  source: DraggableLocation,
  destination: DraggableLocation,
): MoveCrossLaneActionPayload => ({
  from: {
    laneId: source.droppableId,
    index: source.index,
  },
  to: {
    laneId: destination.droppableId,
    index: destination.index,
  },
})

export const getReorderedTicketList = (
  tickets: Ticket[],
  from: number,
  to: number,
) => {
  const result = [...tickets]
  const [movingTicket] = result.splice(from, 1)
  result.splice(to, 0, movingTicket)

  return result
}

export const getMovedTicketLists = (
  source: Ticket[],
  sink: Ticket[],
  from: number,
  to: TicketPosition,
): [Ticket[], Ticket[]] => [
  source.filter((_, index) => index !== from),
  sink
    .slice(0, to.index)
    .concat({
      ...source[from],
      status: getStatusFromString(to.laneId as TicketStatusKeys),
    })
    .concat(sink.slice(to.index)),
]
