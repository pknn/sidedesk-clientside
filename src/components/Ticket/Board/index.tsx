import React from 'react'
import { DragDropContext, DropResult } from 'react-beautiful-dnd'

import { Lane } from 'app/components/Ticket/Lane'
import { TicketStatus } from 'app/types/Ticket'
import { ticketStatusOptions } from 'app/helpers/mockTicket'
import {
  getKeyFromStatus,
  getStatusFromString,
  TicketStatusKeys,
} from 'app/helpers/statusMappers'
import { useAppDispatch, useAppSelector } from 'app/types/Store'
import { BoardState, MoveActionPayload } from 'app/types/Board'
import { actions } from 'app/store/features/boardSlice'

const toLaneComponent = (state: BoardState) =>
  ticketStatusOptions.map((ticketStatus: TicketStatus) => (
    <Lane
      key={ticketStatus}
      tickets={state[getKeyFromStatus(ticketStatus)]}
      laneStatus={ticketStatus}
    />
  ))

export const Board = (): JSX.Element => {
  const boardState = useAppSelector((state) => state.board)
  const dispatch = useAppDispatch()
  const handleDragEnd = (result: DropResult) => {
    const { source, destination } = result
    if (!destination) return
    const payload: MoveActionPayload = {
      from: {
        index: source.index,
        statusLane: getStatusFromString(source.droppableId as TicketStatusKeys),
      },
      to: {
        index: destination.index,
        statusLane: getStatusFromString(
          destination.droppableId as TicketStatusKeys,
        ),
      },
    }
    dispatch(actions.move(payload))
  }

  return (
    <div className="w-full flex justify-between">
      <DragDropContext onDragEnd={handleDragEnd}>
        {toLaneComponent(boardState)}
      </DragDropContext>
    </div>
  )
}
