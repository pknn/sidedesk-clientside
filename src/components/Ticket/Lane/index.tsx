import React from 'react'

import { getMockTickets, ticketStatusOptions } from 'app/helpers/mockTicket'
import { LaneComponent } from './LaneComponent'

export const Lane = (): JSX.Element => {
  const tickets = getMockTickets(50)
  return (
    <div className="w-full flex justify-between">
      {ticketStatusOptions.map((ticketStatus) => (
        <LaneComponent
          key={ticketStatus}
          tickets={tickets.filter((ticket) => ticket.status === ticketStatus)}
          laneStatus={ticketStatus}
        />
      ))}
    </div>
  )
}
