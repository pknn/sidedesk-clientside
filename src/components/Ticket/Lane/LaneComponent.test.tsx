import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { getMockTickets } from 'app/helpers/mockTicket'
import { TicketStatus } from 'app/types/Ticket'

import { Lane } from '.'

const setupLaneComponent = () => {
  const mockTickets = getMockTickets(50).filter(
    (ticket) => ticket.status === TicketStatus.Accepted,
  )
  return shallow(
    <Lane tickets={mockTickets} laneStatus={TicketStatus.Accepted} />,
  )
}

describe('<LaneComponent />', () => {
  it('should render correctly', () => {
    const shallowMountedComponent = setupLaneComponent()

    expect(toJson(shallowMountedComponent)).toMatchSnapshot()
  })
})
