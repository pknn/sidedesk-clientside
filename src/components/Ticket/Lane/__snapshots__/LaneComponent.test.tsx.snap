// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaneComponent /> should render correctly 1`] = `
<styled.div>
  <div
    className="bg-gray-50 rounded shadow-sm h-full relative"
  >
    <div
      className="flex items-center sticky top-0 bg-gray-50 px-2 py-4"
    >
      <div
        className="w-2 h-2 rounded-full mr-2 bg-red-400"
      />
      <span
        className="font-semibold text-gray-500 text-sm italic"
      >
        REJECTED
      </span>
    </div>
    <div
      className="p-2 pt-0"
    >
      <CardComponent
        id={7532}
        key="7532"
        reporterName="Jalen Rutherford"
        title="Repudiandae accusantium et nihil aut in animi numquam quis laborum."
      />
      <CardComponent
        id={74486}
        key="74486"
        reporterName="Genevieve Simonis"
        title="Excepturi vel reiciendis excepturi rem omnis repudiandae odit quis nulla."
      />
      <CardComponent
        id={39594}
        key="39594"
        reporterName="Daphney Harvey"
        title="Earum natus fuga consequatur voluptate recusandae dolor placeat dolore sed."
      />
      <CardComponent
        id={40493}
        key="40493"
        reporterName="Pink Beahan"
        title="Veniam beatae in voluptate numquam quia sit fugit eligendi nihil."
      />
      <CardComponent
        id={17815}
        key="17815"
        reporterName="Alia Bahringer"
        title="Corrupti porro quo fuga suscipit deserunt repellendus dignissimos hic ad."
      />
      <CardComponent
        id={81310}
        key="81310"
        reporterName="Jarvis Kozey"
        title="Accusantium reprehenderit qui alias beatae aut repellat magnam deleniti et."
      />
      <CardComponent
        id={21072}
        key="21072"
        reporterName="Chelsie Gleichner"
        title="Sint nisi recusandae iste consequatur dolores saepe architecto quae omnis."
      />
      <CardComponent
        id={13400}
        key="13400"
        reporterName="Clara Kautzer"
        title="Et eveniet autem consectetur fugiat eum aut aperiam rerum sit."
      />
      <CardComponent
        id={81271}
        key="81271"
        reporterName="Alek Ryan"
        title="Necessitatibus sunt voluptatem sed nam eos ducimus corrupti rerum aperiam."
      />
      <CardComponent
        id={49192}
        key="49192"
        reporterName="Lori Stoltenberg"
        title="Et nesciunt et est expedita iste distinctio consequatur vel blanditiis."
      />
      <CardComponent
        id={22209}
        key="22209"
        reporterName="Nelda Langosh"
        title="Ipsum neque repellendus repudiandae iste provident harum cupiditate hic magni."
      />
      <CardComponent
        id={18215}
        key="18215"
        reporterName="Fausto Erdman"
        title="Sequi earum fugiat nihil similique hic qui eos ut sequi."
      />
      <CardComponent
        id={3496}
        key="3496"
        reporterName="Derek Sipes"
        title="Perferendis architecto et et eos veritatis quo dolores nulla commodi."
      />
      <CardComponent
        id={5661}
        key="5661"
        reporterName="Kira Hegmann"
        title="Eos odit animi voluptatem minus quasi eos corporis quibusdam vel."
      />
      <CardComponent
        id={33630}
        key="33630"
        reporterName="Abby Terry"
        title="Sunt et iusto facilis sit tempore velit repellat ut modi."
      />
      <CardComponent
        id={26500}
        key="26500"
        reporterName="Ethan Schumm"
        title="Qui autem nihil iusto aut vero dolor velit voluptatum tempora."
      />
      <CardComponent
        id={99812}
        key="99812"
        reporterName="Montana Weissnat"
        title="Quam voluptatem error corporis quis illum sed est perferendis tempora."
      />
      <CardComponent
        id={14097}
        key="14097"
        reporterName="Mariana Lind"
        title="Neque quia quia nostrum aperiam quasi ipsa quaerat deserunt dolor."
      />
      <CardComponent
        id={69132}
        key="69132"
        reporterName="Stacey Stamm"
        title="Laborum et est sit occaecati accusantium autem nobis ea ut."
      />
      <CardComponent
        id={80515}
        key="80515"
        reporterName="Jeffery Cole"
        title="Id ducimus quia esse mollitia autem nihil repellendus velit aliquam."
      />
      <CardComponent
        id={1466}
        key="1466"
        reporterName="Ivy Lynch"
        title="Consectetur facere in quis a sapiente omnis nulla enim dolores."
      />
      <CardComponent
        id={47020}
        key="47020"
        reporterName="Pete Renner"
        title="Dignissimos tenetur eum aut corporis animi illo aut impedit deleniti."
      />
      <CardComponent
        id={55880}
        key="55880"
        reporterName="Martin Ernser"
        title="Pariatur distinctio fuga minima ipsum rem nesciunt non ipsa officiis."
      />
      <CardComponent
        id={3732}
        key="3732"
        reporterName="Brigitte Berge"
        title="Fugit tempora aut rerum sunt assumenda repellat alias exercitationem eveniet."
      />
      <CardComponent
        id={98766}
        key="98766"
        reporterName="Judson Cole"
        title="Et aut itaque laborum qui necessitatibus hic sed doloribus consequatur."
      />
      <CardComponent
        id={78932}
        key="78932"
        reporterName="Bethel Pacocha"
        title="Aliquid expedita fugiat minus repudiandae ab aut molestiae sit ut."
      />
      <CardComponent
        id={96638}
        key="96638"
        reporterName="Pink Jaskolski"
        title="Eligendi ex blanditiis sequi dolores deserunt suscipit sapiente beatae aperiam."
      />
      <CardComponent
        id={77601}
        key="77601"
        reporterName="Cecile Romaguera"
        title="Tempore eum tenetur ea possimus ab cum doloremque et quisquam."
      />
      <CardComponent
        id={12109}
        key="12109"
        reporterName="Elbert Maggio"
        title="Neque dolore sint ad voluptatem et et ut ratione qui."
      />
      <CardComponent
        id={63114}
        key="63114"
        reporterName="Lester Steuber"
        title="Neque aut eos distinctio dignissimos qui tempore sit ea qui."
      />
      <CardComponent
        id={55754}
        key="55754"
        reporterName="Kieran Conn"
        title="Quisquam repudiandae ut in aut cumque distinctio exercitationem aperiam debitis."
      />
      <CardComponent
        id={7162}
        key="7162"
        reporterName="Lyda Fadel"
        title="Sunt dolores cumque sit est aut voluptate aut sed magnam."
      />
      <CardComponent
        id={96355}
        key="96355"
        reporterName="Santa Mante"
        title="Quia occaecati maxime et id nam ea dignissimos sit ut."
      />
      <CardComponent
        id={52557}
        key="52557"
        reporterName="Evangeline Buckridge"
        title="Adipisci numquam quo numquam aspernatur sed quia velit possimus harum."
      />
      <CardComponent
        id={33329}
        key="33329"
        reporterName="Antonina Stamm"
        title="Culpa et nam sint sit eum quia provident enim id."
      />
      <CardComponent
        id={40765}
        key="40765"
        reporterName="Bonnie Morar"
        title="Dolorum dolor vel omnis rerum deserunt placeat necessitatibus vel qui."
      />
      <CardComponent
        id={97593}
        key="97593"
        reporterName="Lorna Donnelly"
        title="Mollitia vel voluptates odit ea ut quos in ducimus optio."
      />
      <CardComponent
        id={10982}
        key="10982"
        reporterName="Brando Goldner"
        title="Aliquam numquam nam possimus est molestiae ratione doloribus aut molestias."
      />
      <CardComponent
        id={68400}
        key="68400"
        reporterName="Felicity Shanahan"
        title="Iure exercitationem cupiditate et dolor nihil eos doloribus est enim."
      />
      <CardComponent
        id={68944}
        key="68944"
        reporterName="Myrna Schmitt"
        title="Sit molestias voluptatem repellendus quis voluptatem eos accusamus laudantium maxime."
      />
      <CardComponent
        id={97809}
        key="97809"
        reporterName="Evans Kuhn"
        title="Reprehenderit ab molestiae autem aspernatur similique quo qui explicabo repudiandae."
      />
      <CardComponent
        id={43974}
        key="43974"
        reporterName="Rigoberto Pfeffer"
        title="Harum laboriosam reiciendis saepe natus doloribus itaque sit et quos."
      />
      <CardComponent
        id={53871}
        key="53871"
        reporterName="Janie Feeney"
        title="Corrupti perferendis sint porro et laboriosam excepturi illum minus omnis."
      />
      <CardComponent
        id={10521}
        key="10521"
        reporterName="Marge Tillman"
        title="Qui omnis sequi provident enim vel eos quos sequi magnam."
      />
      <CardComponent
        id={68902}
        key="68902"
        reporterName="Joanny Balistreri"
        title="Ut voluptas non enim neque cum repellat libero sequi modi."
      />
      <CardComponent
        id={23986}
        key="23986"
        reporterName="Veda Ferry"
        title="Fugiat modi dignissimos et quam qui aut sunt a modi."
      />
      <CardComponent
        id={99959}
        key="99959"
        reporterName="Stephanie Kassulke"
        title="Beatae sint ducimus fuga et mollitia nisi ullam repellat quas."
      />
      <CardComponent
        id={34254}
        key="34254"
        reporterName="Ruthie Trantow"
        title="At quo voluptates delectus id est sed veniam quia iusto."
      />
      <CardComponent
        id={41404}
        key="41404"
        reporterName="Sean Altenwerth"
        title="Illum expedita dicta nulla est quo dolores vel quo quis."
      />
      <CardComponent
        id={35112}
        key="35112"
        reporterName="Golda Block"
        title="Perspiciatis doloribus ratione omnis maxime unde ea in sed et."
      />
    </div>
  </div>
</styled.div>
`;
